spec/lib/msf/debug_spec.rb

```diff
@@ -0,0 +1,700 @@
+
+        ```
+        0      Command 1
+        1      Command 2
+        2      Command 3
+        3      Command 4
+        4      Command 5
+        5      Command 6
+        6      Command 7
+        7      Command 8
+        8      Command 9
+        9      Command 10
+        ```
+
+        </details>
+
+
+      E_LOG
+
+      expect(subject.history).to eql(error_log_output)
+    end
+
+    it "correctly retrieves & parses a command history larger than the command total" do
+      Readline::HISTORY = Array.new(15) { |i| "Command #{i+1}"}
+      error_log_output = <<~E_LOG
+        ##  %grnHistory%clr
+        The following commands were ran before this issue occurred:
+        <details>
+        <summary>Collapse</summary>
+
+        ```
+        0      Command 1
+        1      Command 2
+        2      Command 3
+        3      Command 4
+        4      Command 5
+        5      Command 6
+        6      Command 7
+        7      Command 8
+        8      Command 9
+        9      Command 10
+        10     Command 11
+        11     Command 12
+        12     Command 13
+        13     Command 14
+        14     Command 15
+        ```
+
+        </details>
+
+
+      E_LOG
+
+      expect(subject.history).to eql(error_log_output)
+    end
+
+    it "correctly retrieves & parses an empty config file & datastore" do
+      allow(::Msf::Config).to receive(:config_file).and_return(File.join(FILE_FIXTURES_PATH, 'debug', 'config_files', 'empty.ini'))
+
+      framework = double('framework')
+      expect(framework).to receive(:datastore).and_return({})
+
+
+      driver = double('driver')
+      expect(driver).to receive(:get_config_core).and_return('config_core')
+      expect(driver).to receive(:get_config).and_return({})
+      expect(driver).to receive(:get_config_group).and_return('config_group')
+      expect(driver).to receive(:active_module).and_return(nil)
+
+      expected_output = <<~OUTPUT
+      ##  %grnModule/Datastore%clr
+      The following global/module datastore, & databse setup was configured before the issue occurred:
+      <details>
+      <summary>Collapse</summary>
+
+      ```
+      The local config file is empty, no global variables are set, and there is no active module.
+      ```
+
+      </details>
+
+
+      OUTPUT
+
+      expect(subject.datastore(framework, driver)).to eql(expected_output)
+    end
+
+    it "correctly retrieves & parses a populated global datastore" do
+      allow(::Msf::Config).to receive(:config_file).and_return(File.join(FILE_FIXTURES_PATH, 'debug', 'config_files', 'empty.ini'))
+
+      framework = double('framework')
+      expect(framework).to receive(:datastore).and_return({
+                                                            'key1' => 'val1',
+                                                            'key2' => 'val2',
+                                                            'key3' => 'val3'
+                                                           })
```

Comment:
RSpec has a great way to mock objects:

```ruby
      framework = instance_double(
        ::Msf::Framework,
        datastore: {
          'key1' => 'val1',
          'key2' => 'val2',
          'key3' => 'val3'
        }
      )
```

RSpec will also ensure that if the API changes in the underlying object, it'll fail the test. This also helps catch typos, for instance:

```ruby
      framework = instance_double(
        ::Msf::Framework,
        dataastore: {
          'key1' => 'val1',
          'key2' => 'val2',
          'key3' => 'val3'
        }
      )
```

Will give:

```
  1) Debug command functionality Msf::Ui::Debug correctly retrieves & parses a populated global datastore
     Failure/Error:
          ...

          the Msf::Framework class does not implement the instance method: dataastore
```

Reply:
There's other places in the code that we could swap over to using this pattern too 👀